### DISCUSSION TITLE
Custom parser fails after change to SO 2.3.140

### DISCUSSION THREAD

**Discussion by sleepingbel** [Security-Onion-Solutions/securityonion#8365] [Unsupported Versions] [answered]
Hello all,

This is a fresh install
8 vCPU
24 vRAM
1.5 TB Diskspace
Standalone 2.3.140

For this installation I always used a my own custom parser for the syslog logs. With this installation I chose to change the default syslog parser.

My custom parser for pfsense openvpn is not working anymore, because i can not add the needed  changes in the custom syslog parser.

I used the default parser as basis
[SO Respository Default Parsers](https://github.com/Security-Onion-Solutions/securityonion/blob/master/salt/elasticsearch/files/ingest/syslog)

This is my parser, I only added a extra set and pipeline part for openvpn 

```
{
  "description" : "syslog pipeline",
  "processors" : [
    {
      "dissect": {
        "field": "message",
        "pattern" : "%{message}",
        "on_failure": [ { "drop" : { } } ]
      },
      "remove": {
        "field": [ "type", "agent" ],
        "ignore_failure": true
      }
    }, {
      "grok": {
        "field": "message",
          "patterns": [
            "^<%{INT:syslog.priority:int}>%{TIMESTAMP_ISO8601:syslog.timestamp} +%{IPORHOST:syslog.host} +%{PROG:syslog.program}(?:\\[%{POSINT:syslog.pid:int}\\])?: %{GREEDYDATA:real_message}$",

            "^<%{INT:syslog.priority}>%{DATA:syslog.timestamp} %{WORD:source.application}(\\[%{DATA:pid}\\])?: %{GREEDYDATA:real_message}$",

            "^%{SYSLOGTIMESTAMP:syslog.timestamp} %{SYSLOGHOST:syslog.host} %{SYSLOGPROG:syslog.program}: CEF:0\\|%{DATA:vendor}\\|%{DATA:product}\\|%{GREEDYDATA:message2}$",

            "^%{SYSLOGTIMESTAMP:syslog.timestamp} %{SYSLOGHOST:syslog.host} %{SYSLOGPROG:syslog.program}: %{GREEDYDATA:real_message}$"

          ],
          "ignore_failure": true
      }
    },
    {
      "convert" : {
        "if": "ctx?.syslog?.priority != null",
        "field" : "syslog.priority",
        "type": "integer"
      }
    },
    {
      "script": {
        "description": "Map syslog priority into facility and level",
        "lang": "painless",
        "params" : {
          "level": [
            "emerg",
            "alert",
            "crit",
            "err",
            "warn",
            "notice",
            "info",
            "debug"
           ],
           "facility" : [
             "kern",
             "user",
             "mail",
             "daemon",
             "auth",
             "syslog",
             "lpr",
             "news",
             "uucp",
             "cron",
             "authpriv",
             "ftp",
             "ntp",
             "security",
             "console",
             "solaris-cron",
             "local0",
             "local1",
             "local2",
             "local3",
             "local4",
             "local5",
             "local6",
             "local7"
            ]
        },
        "source": "if (ctx['syslog'] != null && ctx['syslog']['priority'] != null) { int p = ctx['syslog']['priority']; int f = p / 8; int l = p - (f * 8); ctx['syslog']['facility_label'] = [ : ]; ctx['syslog']['severity_label'] = [ : ]$

      }
    }, {
      "set": {
        "if": "ctx.syslog?.host != null",
        "field": "host.name",
        "value": "{{ syslog.host }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.syslog?.program != null",
        "field": "process.name",
        "value": "{{ syslog.program }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.syslog?.pid != null",
        "field": "process.id",
        "value": "{{ syslog.pid }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.source?.application == 'filterlog'",
        "field": "dataset",
        "value": "firewall",
        "ignore_failure": true
      }
    }, { 
      "set": {
        "if": "ctx.source?.application == 'openvpn'",
        "field": "dataset",
        "value": "vpn", 
        "ignore_failure": true  
      }
    }, {
      "set": {
        "if": "ctx.vendor != null",
        "field": "module",
        "value": "{{ vendor }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "if": "ctx.product != null",
        "field": "dataset",
        "value": "{{ product }}",
        "ignore_failure": true
      }
    }, {
      "set": {
        "field": "ingest.timestamp",
        "value": "{{ @timestamp }}"
      }
    }, {
      "date": {
        "if": "ctx.syslog?.timestamp != null",
        "field": "syslog.timestamp",
        "target_field": "@timestamp",
        "formats": ["MMM  d HH:mm:ss", "MMM dd HH:mm:ss", "ISO8601", "UNIX"],
        "ignore_failure": true
      }
    }, {
      "remove": {
        "field": ["pid", "program"],
        "ignore_missing": true,
        "ignore_failure": true
      }
    }, {
      "pipeline": {
        "if": "ctx.vendor != null && ctx.product != null",
        "name": "{{ vendor }}.{{ product }}",
        "ignore_failure": true
      }
    }, {
      "pipeline": {
        "if": "ctx.dataset == 'firewall'",
        "name": "filterlog",
        "ignore_failure": true
      }
    }, { 
      "pipeline": { 
        "if": "ctx.dataset == 'vpn'", 
        "name": "vpn", 
        "ignore_failure": true 
      } 
    }, {
      "pipeline": { "name": "common" }
    }
  ]
}
```

This is the error that appears in Loading pipelines

`
                  syslog
                  {"error":{"root_cause":[{"type":"x_content_parse_exception","reason":"[1:2214] Unexpected character ('s' (code 115)): was expecting comma to separate Object entries\n at [Source: (byte[])\"{ \"description\" : \"syslog pipeline\", \"processors\" : [ { \"dissect\": { \"field\": \"message\", \"pattern\" : \"%{message}\", \"on_failure\": [ { \"drop\" : { } } ] }, \"remove\": { \"field\": [ \"type\", \"agent\" ], \"ignore_failure\": true } }, { \"grok\": { \"field\": \"message\", \"patterns\": [ \"^<%{INT:syslog.priority:int}>%{TIMESTAMP_ISO8601:syslog.timestamp} +%{IPORHOST:syslog.host} +%{PROG:syslog.program}(?:\\\\[%{POSINT:syslo\"[truncated 3815 bytes]; line: 1, column: 2214]"}],"type":"x_content_parse_exception","reason":"[1:2214] Unexpected character ('s' (code 115)): was expecting comma to separate Object entries\n at [Source: (byte[])\"{ \"description\" : \"syslog pipeline\", \"processors\" : [ { \"dissect\": { \"field\": \"message\", \"pattern\" : \"%{message}\", \"on_failure\": [ { \"drop\" : { } } ] }, \"remove\": { \"field\": [ \"type\", \"agent\" ], \"ignore_failure\": true } }, { \"grok\": { \"field\": \"message\", \"patterns\": [ \"^<%{INT:syslog.priority:int}>%{TIMESTAMP_ISO8601:syslog.timestamp} +%{IPORHOST:syslog.host} +%{PROG:syslog.program}(?:\\\\[%{POSINT:syslo\"[truncated 3815 bytes]; line: 1, column: 2214]","caused_by":{"type":"json_parse_exception","reason":"Unexpected character ('s' (code 115)): was expecting comma to separate Object entries\n at [Source: (byte[])\"{ \"description\" : \"syslog pipeline\", \"processors\" : [ { \"dissect\": { \"field\": \"message\", \"pattern\" : \"%{message}\", \"on_failure\": [ { \"drop\" : { } } ] }, \"remove\": { \"field\": [ \"type\", \"agent\" ], \"ignore_failure\": true } }, { \"grok\": { \"field\": \"message\", \"patterns\": [ \"^<%{INT:syslog.priority:int}>%{TIMESTAMP_ISO8601:syslog.timestamp} +%{IPORHOST:syslog.host} +%{PROG:syslog.program}(?:\\\\[%{POSINT:syslo\"[truncated 3815 bytes]; line: 1, column: 2214]"}},"status":400}
`  

Can someone help me with this?
Regards
Bart

**Comment by btdewey** [comment, marked as answer]
Putting your parser in a yaml checker (https://yamlchecker.com/) comes back with the following error:
```
missed comma between flow collection entries (80:8)

 77 | 
 78 |       }
 79 |     }, {
 80 |       "set": {
-------------^
 81 |         "if": "ctx.syslog?.host != null",
 82 |         "field": "host.name",
```
This seems to match the error thrown by SO.  This seems to indicate a missing special character higher up in the yaml file.  Taking a deeper look there appears to be a missing trailing double quote on line 78.  The if statement isn't ended properly.  Adding a double quote at the end of this line solves the yaml error and should fix your issue.

`        "source": "if (ctx['syslog'] != null && ctx['syslog']['priority'] != null) { int p = ctx['syslog']['priority']; int f = p / 8; int l = p - (f * 8); ctx['syslog']['facility_label'] = [ : ]; ctx['syslog']['severity_label'] = [ : ]$`

**Reply by sleepingbel** [reply]
Thx, I think the double quote was lost in copying

**Reply by sleepingbel** [reply]
This gives still a error when I restart elasticsearch

` syslog
                  {"error":{"root_cause":[{"type":"script_exception","reason":"compile error","processor_type":"script","script_stack":["... 'severity_label'] = [ : ]$"," ^---- HERE"],"script":"if (ctx['syslog'] != null && ctx['syslog']['priority'] != null) { int p = ctx['syslog']['priority']; int f = p / 8; int l = p - (f beats.common common common.nids dns.tld ecs filterlog http.status import.wel kratos logscan.alert osquery.live_query osquery.normalize osquery.query_result ossec rita.beacon rita.connection rita.dns strelka.file suricata.alert suricata.common suricata.dhcp suricata.dnp3 suricata.dns suricata.fileinfo suricata.flow suricata.ftp suricata.ftp_data suricata.http suricata.ikev2 suricata.krb5 suricata.nfs suricata.rdp suricata.sip suricata.smb suricata.smtp suricata.snmp suricata.ssh suricata.tftp suricata.tls syslog sysmon vpn win.eventlogs zeek.common zeek.common_ssl zeek.conn zeek.dce_rpc zeek.dhcp zeek.dnp3 zeek.dns zeek.dpd zeek.files zeek.ftp zeek.http zeek.intel zeek.irc zeek.kerberos zeek.modbus zeek.mysql zeek.notice zeek.ntlm zeek.pe zeek.radius zeek.rdp zeek.rfb zeek.signatures zeek.sip zeek.smb_files zeek.smb_mapping zeek.smtp zeek.snmp zeek.socks zeek.software zeek.ssh zeek.ssl zeek.syslog zeek.tunnel zeek.tunnels zeek.weird zeek.x509 8); ctx['syslog']['facility_label'] = [ : ]; ctx['syslog']['severity_label'] = [ : ]$","lang":"painless","position":{"offset":217,"start":192,"end":218}}],"type":"script_exception","reason":"compile error","processor_type":"script","script_stack":["... 'severity_label'] = [ : ]$"," ^---- HERE"],"script":"if (ctx['syslog'] != null && ctx['syslog']['priority'] != null) { int p = ctx['syslog']['priority']; int f = p / 8; int l = p - (f beats.common common common.nids dns.tld ecs filterlog http.status import.wel kratos logscan.alert osquery.live_query osquery.normalize osquery.query_result ossec rita.beacon rita.connection rita.dns strelka.file suricata.alert suricata.common suricata.dhcp suricata.dnp3 suricata.dns suricata.fileinfo suricata.flow suricata.ftp suricata.ftp_data suricata.http suricata.ikev2 suricata.krb5 suricata.nfs suricata.rdp suricata.sip suricata.smb suricata.smtp suricata.snmp suricata.ssh suricata.tftp suricata.tls syslog sysmon vpn win.eventlogs zeek.common zeek.common_ssl zeek.conn zeek.dce_rpc zeek.dhcp zeek.dnp3 zeek.dns zeek.dpd zeek.files zeek.ftp zeek.http zeek.intel zeek.irc zeek.kerberos zeek.modbus zeek.mysql zeek.notice zeek.ntlm zeek.pe zeek.radius zeek.rdp zeek.rfb zeek.signatures zeek.sip zeek.smb_files zeek.smb_mapping zeek.smtp zeek.snmp zeek.socks zeek.software zeek.ssh zeek.ssl zeek.syslog zeek.tunnel zeek.tunnels zeek.weird zeek.x509 8); ctx['syslog']['facility_label'] = [ : ]; ctx['syslog']['severity_label'] = [ : ]$","lang":"painless","position":{"offset":217,"start":192,"end":218},"caused_by":{"type":"illegal_argument_exception","reason":"invalid sequence of tokens near ['$'].","caused_by":{"type":"no_viable_alt_exception","reason":null}}},"status":400}
`

**Reply by sleepingbel** [reply]
I found the problem, human error, I have to check the code carefully otherwise it will never work
